src/serializer/ResidualHeapVisitor.js

```diff
@@ -400,26 +394,66 @@ export class ResidualHeapVisitor {
     }
   }
 
-  visitValueSet(val: ObjectValue): void {
-    let kind = val.getKind();
+  visitValueWeakMap(val: ObjectValue): void {
+    invariant(val.getKind() === "WeakMap");
+    let entries = val.$WeakMapData;
 
-    let entries;
-    if (kind === "Set") {
-      entries = val.$SetData;
-    } else {
-      invariant(kind === "WeakSet");
-      entries = val.$WeakSetData;
-    }
     invariant(entries !== undefined);
     let len = entries.length;
 
+    for (let i = 0; i < len; i++) {
+      this._withUnrelatedScope(this.scope, () => {
+        let entry = entries[i];
+        let key = entry.$Key;
+        let value = entry.$Value;
+
+        if (key !== undefined && value !== undefined) {
+          if (this.values.has(key)) {
+            this.visitValue(key);
+            this.visitValue(value);
+          } else {
+            this._withUnrelatedScope(this.scope, () => {
+              if (key !== undefined && value !== undefined && this.values.has(key)) {
+                this.visitValue(key);
+                this.visitValue(value);
+              }
+            });
+          }
+        }
+      });
+    }
+  }
+
+  visitValueSet(val: ObjectValue): void {
+    invariant(val.getKind() === "Set");
+
+    let entries = val.$SetData;
+    invariant(entries !== undefined);
+
+    let len = entries.length;
     for (let i = 0; i < len; i++) {
       let entry = entries[i];
       if (entry === undefined) continue;
       this.visitValue(entry);
     }
   }
 
+  visitValueWeakSet(val: ObjectValue): void {
+    invariant(val.getKind() === "WeakSet");
+
+    let entries = val.$WeakSetData;
+    invariant(entries !== undefined);
+
+    this._withUnrelatedScope(this.scope, () => {
+      let len = entries.length;
+      for (let i = 0; i < len; i++) {
+        let entry = entries[i];
+        if (entry === undefined || !this.values.has(entry)) continue;
```

Comment:
Even for a set, values might be visited for some other reason later on. So each entry here should be wrapped in withUnrelatedScope calls that keep re-issuing withUnrelatedScope calls when the value hasn't been visited yet.